### ISSUE TITLE
error undefined method 'version' for #<Array:0x0000000b9db290>

### ISSUE THREAD

**Issue by Stromweld** [chef-cookbooks/chef_client_updater#41]
### Cookbook version
2.0.3

### Chef-client version
13.1.31

### Platform Details
Windows Server 2012R2 running on AWS Cloud

### Scenario:
normal chef-client run

### Steps to Reproduce:
run chef-client with chef_client_updater in runlist

### Expected Result:
Chef to finish convergence

### Actual Result:
error on several but not all of my windows servers.
````
Recipe: chef_client_updater::default
  * chef_client_updater[update chef-client] action update[2017-06-28T10:00:52-05:00] WARN: NO c:/opscode/chef.upgrade DI
R TO ROLL BACK TO!


    ================================================================================
    Error executing action `update` on resource 'chef_client_updater[update chef-client]'
    ================================================================================

    NoMethodError
    -------------
    undefined method `version' for #<Array:0x0000000b9db290>

    Resource Declaration:
    ---------------------
    # In C:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb

     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25:   download_url_override node['chef_client_updater']['download_url_override'] if node['chef_client_updater']['do
wnload_url_override']
     26:   checksum node['chef_client_updater']['checksum'] if node['chef_client_updater']['checksum']
     27: end

    Compiled Resource:
    ------------------
    # Declared in C:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

    chef_client_updater("update chef-client") do
      action [:update]
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "latest"
      prevent_downgrade true
      post_install_action "exec"
    end

    System Info:
    ------------
    chef_version=13.1.31
    platform=windows
    platform_version=6.3.9600
    ruby=ruby 2.4.1p111 (2017-03-22 revision 58053) [x64-mingw32]
    program_name=C:/opscode/chef/bin/chef-client
    executable=C:/opscode/chef/bin/chef-client


Running handlers:
[2017-06-28T10:00:52-05:00] ERROR: Running exception handlers
Running handlers complete
[2017-06-28T10:00:52-05:00] ERROR: Exception handlers complete
Chef Client failed. 1 resources updated in 35 seconds
[2017-06-28T10:00:52-05:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2017-06-28T10:00:52-05:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2017-06-28T10:00:52-05:00] FATAL: NoMethodError: chef_client_updater[update chef-client] (chef_client_updater::default
line 20) had an error: NoMethodError: undefined method `version' for #<Array:0x0000000b9db290>
````
chef-stacktrace.out:
````
Generated at 2017-06-28 10:00:52 -0500
NoMethodError: chef_client_updater[update chef-client] (chef_client_updater::default line 20) had an error: NoMethodError: undefined method `version' for #<Array:0x0000000b9db290>
C:/chef/cache/cookbooks/chef_client_updater/providers/default.rb:112:in `update_necessary?'
C:/chef/cache/cookbooks/chef_client_updater/providers/default.rb:228:in `block in class_from_file'
(eval):2:in `block in action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:211:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:211:in `compile_and_converge_action'
(eval):2:in `action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:171:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource.rb:591:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:715:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:710:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:710:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:749:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:286:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:273:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:249:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:232:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:450:in `loop'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:450:in `interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:434:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:59:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:58:in `load'
C:/opscode/chef/bin/chef-client:58:in `<main>'

>>>> Caused by NoMethodError: undefined method `version' for #<Array:0x0000000b9db290>
C:/chef/cache/cookbooks/chef_client_updater/providers/default.rb:112:in `update_necessary?'
C:/chef/cache/cookbooks/chef_client_updater/providers/default.rb:228:in `block in class_from_file'
(eval):2:in `block in action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:211:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:211:in `compile_and_converge_action'
(eval):2:in `action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/provider.rb:171:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource.rb:591:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:715:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:710:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:710:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:749:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/client.rb:286:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:273:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:249:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:232:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:450:in `loop'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:450:in `interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application/client.rb:434:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/lib/chef/application.rb:59:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.1.31-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:58:in `load'
C:/opscode/chef/bin/chef-client:58:in `<main>'
````

**Comment by lamont-granquist** [comment]
@wrightp it looks like we're getting scalars or arrays of scalars back from the mixlib-install API here, depending on conditions, which should probably be fixed in mixlib-install so it returns either an array or a scalar?

**Comment by wrightp** [comment]
`#artifact_info` will return a scalar if an exact artifact is found based on the query. If multiple artifacts match the query it will return an array of scalars for all matching artifacts. The team has had several discussions on whether to always return an array or to behave as it does now. If you have input please share.

**Comment by lamont-granquist** [comment]
it should be always array.

i'd actually state more strongly that ruby return types should be considered like the /should/ be strongly typed so they're predictable.  also some of the positively *worst* bugs i've seen in the chef client codebase have come from having instance variables like `@url` duck-typing and being settable to either a String or a URI -- which is slick and works fine until it all goes horribly wrong and become nearly impossible for smart people to reason about the behavior.  i'm actually somewhat skeptical as well about automatic coercion on the input side of methods, and the stuff like `#to_s`'ing an input parameter so you can use strings, or symbols or anything that stringifies is convenient but also can be very annoying down the road.  sometimes its better to be strict and only accept one or the other and force the caller to do the explicit coercion (its possible that most of the time i've seen that go badly is the prior example where state gets stored which can be symbol, a string, or any other object that `#to_s`es without explicit coercion and then down the road you buy yourself a heap of pain when someone feeds you a String when conventionally everyone has been using symbols and some consumer of the state explodes because you didn't understand the contract on the state was so loose).

TL;DR:  strongly typed languages have some very good points.

**Comment by wrightp** [comment]
Thanks. I concur. I'll propose this for 4.0.

**Comment by lamont-granquist** [comment]
its easy enough to coerce it to an array here as well, so i can fix this bug and get this issue closed

**Comment by lamont-granquist** [comment]
so @wrightp do i want .first or .last?  i'm assuming .first?

```
Mixlib::Versioning.parse(Array.new(mixlib_install.artifact_info).first.version)
```

**Comment by wrightp** [comment]
I have always used first in this case. All artifacts will be of the same version

**Comment by lamont-granquist** [comment]
ah i get it, cool.

**Comment by wrightp** [comment]
To elaborate... The scalar value will be returned based on platform, platform version, architecture, product, product version, channel.  The array will return all available platform permutations of channel, product, product versions if no platform information is provided to the constructor.

**Comment by Stromweld** [comment]
Is 3.x going to be released soon with this fix in it?
